### ISSUE TITLE
Use cppcheck as a continuous integration check?

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#3108]
@drwells -- you ran cppcheck recently, and it found a number of issues that you fixed. Would it make sense to use this as an additional check in the github continuous integration steps? How long does it run on our code base, and did it find issues that you decided not to fix?

**Comment by Rombur** [comment]
In my experience, cppcheck produces a lot of false positives. We used to run cppcheck for every commit but we stopped doing it because you have to go through 40 warnings to find the only one that's relevant.

**Comment by drwells** [comment]
cppcheck has a lot of trouble with some parts of deal.II. One example: it could not figure out how the serialization classes worked and always complained that variables were unused when we used `operator&`. Another example: cppcheck has trouble with the whole `pointer.reset(new Foo())` idiom that we use with shared pointers and thinks that they are memory leaks. Hence I have to agree with @Rombur: the false positive rate is too high for it to be useful with continuous integration.

The length of time cppcheck requires depends on the mode in which it is run. If one just scans the headers or just scans the source files (e.g., no header files are included) then the process takes about two minutes. If one scans the source files and includes headers (e.g., `cppcheck -j4 -I/home/drwells/Applications/dealii-dev/include/ source/*/*.cc`) then things take about half an hour.

I think it would be worthwhile to add 'run cppcheck and examine the results' to a list of steps we use before making a release. cppcheck is a useful program and we should run it at least infrequently.

I didn't actually use the `-I` option last time I used cppcheck: this seems to catch a different set of things. I'll have some more cppcheck related patches soon. Do we have a list anywhere of the things we do before each release?
